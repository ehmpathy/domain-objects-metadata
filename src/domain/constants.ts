import { createIsOfEnum } from 'type-fns';

export enum DomainObjectVariant {
  DOMAIN_OBJECT = 'DomainObject',
  DOMAIN_LITERAL = 'DomainLiteral',
  DOMAIN_ENTITY = 'DomainEntity',
  DOMAIN_EVENT = 'DomainEvent',
}
export const isOfDomainObjectVariant = createIsOfEnum(DomainObjectVariant);
