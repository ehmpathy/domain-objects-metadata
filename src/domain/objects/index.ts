export * from './DomainObjectMetadata';
export * from './DomainObjectMetadataReference';
export * from './DomainObjectProperty';
