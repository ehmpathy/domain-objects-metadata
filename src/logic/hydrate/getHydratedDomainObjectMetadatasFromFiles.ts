import { UnexpectedCodePathError } from '@ehmpathy/error-fns';
import omit from 'lodash.omit';
import { isPresent } from 'type-fns';
import {
  isClassDeclaration,
  isEnumDeclaration,
  isTypeAliasDeclaration,
  SourceFile,
} from 'typescript';

import {
  DomainObjectMetadata,
  DomainObjectReferenceMetadata,
  DomainObjectPropertyMetadata,
  DomainObjectPropertyType,
} from '../../domain';
import { extractAliasMetadataFromAliasDeclaration } from '../extract/extractAliasMetadataFromAliasDeclaration';
import { extractDomainObjectMetadataForDeclarationInFile } from '../extract/extractDomainObjectMetadataForDeclarationInFile';
import { extractEnumMetadataFromEnumDeclaration } from '../extract/extractEnumMetadataFromEnumDeclaration';
import { isAClassDecorationWhichExtendsDomainObject } from '../extract/isAClassDeclarationWhichExtendsDomainObject';

// hydrate property if needed
const ensurePropertyIsHydrated = ({
  domainObjectName,
  propertyName,
  propertyDefinition: definition,
  domainObjectMetadatas,
  enumMetadatas,
  aliasMetadatas,
}: {
  domainObjectName: string;
  propertyName: string;
  propertyDefinition: DomainObjectPropertyMetadata;
  domainObjectMetadatas: DomainObjectMetadata[];
  enumMetadatas: { name: string; options: string[] }[];
  aliasMetadatas: { name: string; primitive: DomainObjectPropertyType }[];
}): DomainObjectPropertyMetadata => {
  // handle the special case of arrays first; arrays are special because their `of` property is a property definition of its own
  if (definition.type === DomainObjectPropertyType.ARRAY)
    return new DomainObjectPropertyMetadata({
      ...definition,
      of: omit(
        ensurePropertyIsHydrated({
          domainObjectName,
          propertyName,
          propertyDefinition: definition.of as DomainObjectPropertyMetadata, // since arrays have nested property as "of",
          domainObjectMetadatas,
          enumMetadatas,
          aliasMetadatas,
        }),
        'name', // make sure to remove the nested name that comes back when we hydrate a reference property
      ),
    });

  // otherwise, if not a REFERENCE property, then its already hydrated
  if (definition.type !== DomainObjectPropertyType.REFERENCE) return definition; // no change

  // since its a REFERENCE property, hydrate it
  const referencedName = definition.of as string; // reference properties define this as `string` before they're hydrated

  // try to see if its referencing a domain object
  const foundReferencedDomainObject = domainObjectMetadatas.find(
    (metadata) => metadata.name === referencedName,
  );
  if (foundReferencedDomainObject)
    return new DomainObjectPropertyMetadata({
      ...definition,
      name: propertyName, // this is not already on the `definition` if it was nested in an array, so make sure we set it
      of: new DomainObjectReferenceMetadata({
        name: foundReferencedDomainObject.name,
        extends: foundReferencedDomainObject.extends,
      }), // only expose the "name" and "extends" on the nested object metadata
    });

  // try to see if its referencing an enum
  const foundReferencedEnum = enumMetadatas.find(
    (metadata) => metadata.name === referencedName,
  );
  if (foundReferencedEnum)
    return new DomainObjectPropertyMetadata({
      ...definition,
      type: DomainObjectPropertyType.ENUM, // update the name to "enum"
      of: foundReferencedEnum.options,
    });

  // try to see if its referencing a type alias
  const foundReferencedTypeAlias = aliasMetadatas.find(
    (metadata) => metadata.name === referencedName,
  );
  if (foundReferencedTypeAlias)
    return new DomainObjectPropertyMetadata({
      ...definition,
      type: foundReferencedTypeAlias.primitive,
    });

  // if this is neither, throw an error since we can't figure out what its referencing
  throw new UnexpectedCodePathError(
    `could not hydrate a property who references another type. no domain object or enum found with that referenced name. see \`${domainObjectName}.${propertyName}: ${referencedName}\``,
  );
};

export const getHydratedDomainObjectMetadatasFromFiles = (
  files: readonly SourceFile[],
) => {
  // lookup all the unhydrated metadatas for all domain object declarations
  const domainObjectMetadatas = files
    .map((file) => {
      const domainObjectDeclarations = file.statements
        .filter(isClassDeclaration)
        .filter(isAClassDecorationWhichExtendsDomainObject);
      return domainObjectDeclarations.map((classDeclaration) =>
        extractDomainObjectMetadataForDeclarationInFile({
          classDeclaration,
          file,
        }),
      );
    })
    .flat();

  // lookup all the enum metadatas for all of the enum declarations
  const enumMetadatas = files
    .map((file) => {
      const enumDeclarations = file.statements.filter(isEnumDeclaration);
      return enumDeclarations.map(extractEnumMetadataFromEnumDeclaration);
    })
    .flat();

  // lookup all the typealias metadatas for all of the typealiases referenced by our domain object declarations
  const referencedAliasNames: string[] = domainObjectMetadatas
    .map((dobjMetadata) =>
      Object.values(dobjMetadata.properties).map((propertyMetadata) =>
        typeof propertyMetadata.of === 'string'
          ? // can only be an alias if "of" is type of string
            propertyMetadata.of
          : null,
      ),
    )
    .flat()
    .filter(isPresent);
  const aliasMetadatas = files
    .map((file) => {
      const aliasDeclarationsAll = file.statements.filter(
        isTypeAliasDeclaration,
      );
      const aliasDeclarationsWeCareAbout = aliasDeclarationsAll.filter(
        (declaration) => referencedAliasNames.includes(declaration.name.text),
      );
      return aliasDeclarationsWeCareAbout.map(
        extractAliasMetadataFromAliasDeclaration,
      );
    })
    .flat();

  // hydrate every type `reference` property in `domainObjectMetadata`
  const hydratedDomainObjectMetadatas = domainObjectMetadatas.map(
    (metadata) => {
      return new DomainObjectMetadata({
        ...metadata,
        properties: Object.fromEntries(
          Object.entries(metadata.properties).map(([name, definition]) => {
            return [
              name,
              ensurePropertyIsHydrated({
                domainObjectName: metadata.name,
                propertyName: name,
                propertyDefinition: definition,
                domainObjectMetadatas,
                enumMetadatas,
                aliasMetadatas,
              }),
            ];
          }),
        ),
      });
    },
  );

  // and return the hydrated object metadatas
  return hydratedDomainObjectMetadatas;
};
