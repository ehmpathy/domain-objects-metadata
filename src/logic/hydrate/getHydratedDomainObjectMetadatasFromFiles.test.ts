import ts from 'typescript';

import { DomainObjectPropertyType } from '../../domain';
import { getHydratedDomainObjectMetadatasFromFiles } from './getHydratedDomainObjectMetadatasFromFiles';

describe('getHydratedDomainObjectMetadatasFromFiles', () => {
  it('should return metadata from files w/ no need for hydration', () => {
    const program = ts.createProgram(
      [`${__dirname}/../__test_assets__/Address.ts`],
      {},
    );
    const files = program
      .getSourceFiles()
      .filter((file) => !file.fileName.includes('/node_modules/')); // skip the node modules
    const metadatas = getHydratedDomainObjectMetadatasFromFiles(files);
    // console.log(JSON.stringify(metadatas, null, 2));
    expect(metadatas).toMatchSnapshot();
  });
  it('should return metadata from files w/ has lots of imports and needs hydration of nested domain objects', () => {
    const program = ts.createProgram(
      [`${__dirname}/../__test_assets__/Order.ts`],
      {},
    );
    const files = program
      .getSourceFiles()
      .filter((file) => !file.fileName.includes('/node_modules/')); // skip the node modules
    const metadatas = getHydratedDomainObjectMetadatasFromFiles(files);
    // console.log(JSON.stringify(metadatas, null, 2));

    // check that it has everything we expected
    expect(metadatas.length).toEqual(3); // shuld have got both the source and the references
    expect(metadatas[2]?.name).toEqual('Order');
    expect(metadatas[2]?.properties.destination).toMatchObject({
      type: DomainObjectPropertyType.REFERENCE,
      of: { name: 'Address' },
    }); // nested metadata
    expect(metadatas[2]?.properties.items).toMatchObject({
      type: DomainObjectPropertyType.ARRAY,
      of: { type: DomainObjectPropertyType.REFERENCE, of: { name: 'Item' } }, // nested metadata in an array
    });

    // ensure: the nested referenced metadata should not include the full reference. (i.e., no properties / decorations on it)
    expect(metadatas[2]?.properties.destination?.of).not.toHaveProperty(
      'properties',
    );
    expect(metadatas[2]?.properties.destination?.of).not.toHaveProperty(
      'decorations',
    );

    // log an example
    expect(metadatas).toMatchSnapshot();
  });
  it('should return metadata from files which needs hydration of nested enum', () => {
    const program = ts.createProgram(
      [`${__dirname}/../__test_assets__/Payment.ts`],
      {},
    );
    const files = program
      .getSourceFiles()
      .filter((file) => !file.fileName.includes('/node_modules/')); // skip the node modules
    const metadatas = getHydratedDomainObjectMetadatasFromFiles(files);
    // console.log(JSON.stringify(metadatas, null, 2));

    expect(metadatas[0]?.properties.status?.type).toEqual(
      DomainObjectPropertyType.ENUM,
    );
    expect(metadatas[0]?.properties.status?.of).toEqual([
      'PENDING',
      'COMPLETED',
      'CANCELED',
    ]);

    expect(metadatas).toMatchSnapshot();
  });
  it('should return metadata from files which needs hydration of nested type alias', () => {
    const program = ts.createProgram(
      [`${__dirname}/../__test_assets__/PerformanceReport.ts`],
      {},
    );
    const files = program
      .getSourceFiles()
      .filter((file) => !file.fileName.includes('/node_modules/')); // skip the node modules
    const metadatas = getHydratedDomainObjectMetadatasFromFiles(files);
    console.log(JSON.stringify(metadatas, null, 2));

    expect(metadatas[0]?.properties.onDate?.type).toEqual(
      DomainObjectPropertyType.STRING,
    );
    expect(metadatas).toMatchSnapshot();
  });
  it('should return metadata from files which needs hydration of nested enum array', () => {
    const program = ts.createProgram(
      [`${__dirname}/../__test_assets__/AgentRole.ts`],
      {},
    );
    const files = program
      .getSourceFiles()
      .filter((file) => !file.fileName.includes('/node_modules/')); // skip the node modules
    const metadatas = getHydratedDomainObjectMetadatasFromFiles(files);

    expect(metadatas[0]?.properties.permissions?.type).toEqual(
      DomainObjectPropertyType.ARRAY,
    );
    expect((metadatas[0]?.properties.permissions?.of as any).type).toEqual(
      DomainObjectPropertyType.ENUM,
    );
    expect(metadatas).toMatchSnapshot();
  });
});
