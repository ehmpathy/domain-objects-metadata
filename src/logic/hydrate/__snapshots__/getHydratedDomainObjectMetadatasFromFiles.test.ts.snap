// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHydratedDomainObjectMetadatasFromFiles should return metadata from files w/ has lots of imports and needs hydration of nested domain objects 1`] = `
[
  DomainObjectMetadata {
    "decorations": {
      "unique": null,
      "updatable": null,
    },
    "extends": "DomainLiteral",
    "name": "Address",
    "properties": {
      "city": DomainObjectPropertyMetadata {
        "name": "city",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "postal": DomainObjectPropertyMetadata {
        "name": "postal",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "state": DomainObjectPropertyMetadata {
        "name": "state",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "street": DomainObjectPropertyMetadata {
        "name": "street",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "suite": DomainObjectPropertyMetadata {
        "name": "suite",
        "nullable": true,
        "required": true,
        "type": "STRING",
      },
    },
  },
  DomainObjectMetadata {
    "decorations": {
      "unique": [
        "upc",
      ],
      "updatable": [
        "colors",
        "quantity",
        "description",
      ],
    },
    "extends": "DomainEntity",
    "name": "Item",
    "properties": {
      "colors": DomainObjectPropertyMetadata {
        "name": "colors",
        "nullable": false,
        "of": {
          "nullable": false,
          "required": true,
          "type": "STRING",
        },
        "required": true,
        "type": "ARRAY",
      },
      "description": DomainObjectPropertyMetadata {
        "name": "description",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "price": DomainObjectPropertyMetadata {
        "name": "price",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
      "upc": DomainObjectPropertyMetadata {
        "name": "upc",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
    },
  },
  DomainObjectMetadata {
    "decorations": {
      "unique": [],
      "updatable": [],
    },
    "extends": "DomainEntity",
    "name": "Order",
    "properties": {
      "destination": DomainObjectPropertyMetadata {
        "name": "destination",
        "nullable": false,
        "of": DomainObjectReferenceMetadata {
          "extends": "DomainLiteral",
          "name": "Address",
        },
        "required": true,
        "type": "REFERENCE",
      },
      "id": DomainObjectPropertyMetadata {
        "name": "id",
        "nullable": false,
        "required": false,
        "type": "NUMBER",
      },
      "items": DomainObjectPropertyMetadata {
        "name": "items",
        "nullable": false,
        "of": {
          "nullable": false,
          "of": DomainObjectReferenceMetadata {
            "extends": "DomainEntity",
            "name": "Item",
          },
          "required": true,
          "type": "REFERENCE",
        },
        "required": true,
        "type": "ARRAY",
      },
      "status": DomainObjectPropertyMetadata {
        "name": "status",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "uuid": DomainObjectPropertyMetadata {
        "name": "uuid",
        "nullable": false,
        "required": false,
        "type": "STRING",
      },
    },
  },
]
`;

exports[`getHydratedDomainObjectMetadatasFromFiles should return metadata from files w/ no need for hydration 1`] = `
[
  DomainObjectMetadata {
    "decorations": {
      "unique": null,
      "updatable": null,
    },
    "extends": "DomainLiteral",
    "name": "Address",
    "properties": {
      "city": DomainObjectPropertyMetadata {
        "name": "city",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "postal": DomainObjectPropertyMetadata {
        "name": "postal",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "state": DomainObjectPropertyMetadata {
        "name": "state",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "street": DomainObjectPropertyMetadata {
        "name": "street",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "suite": DomainObjectPropertyMetadata {
        "name": "suite",
        "nullable": true,
        "required": true,
        "type": "STRING",
      },
    },
  },
]
`;

exports[`getHydratedDomainObjectMetadatasFromFiles should return metadata from files which needs hydration of nested enum 1`] = `
[
  DomainObjectMetadata {
    "decorations": {
      "unique": [
        "externalId",
      ],
      "updatable": [
        "status",
        "amount",
        "currency",
      ],
    },
    "extends": "DomainEntity",
    "name": "Payment",
    "properties": {
      "amount": DomainObjectPropertyMetadata {
        "name": "amount",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
      "currency": DomainObjectPropertyMetadata {
        "name": "currency",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "externalId": DomainObjectPropertyMetadata {
        "name": "externalId",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "status": DomainObjectPropertyMetadata {
        "name": "status",
        "nullable": false,
        "of": [
          "PENDING",
          "COMPLETED",
          "CANCELED",
        ],
        "required": true,
        "type": "ENUM",
      },
    },
  },
]
`;

exports[`getHydratedDomainObjectMetadatasFromFiles should return metadata from files which needs hydration of nested enum array 1`] = `
[
  DomainObjectMetadata {
    "decorations": {
      "unique": [
        "exid",
      ],
      "updatable": [
        "name",
        "description",
        "permissions",
      ],
    },
    "extends": "DomainEntity",
    "name": "AgentRole",
    "properties": {
      "description": DomainObjectPropertyMetadata {
        "name": "description",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "exid": DomainObjectPropertyMetadata {
        "name": "exid",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "name": DomainObjectPropertyMetadata {
        "name": "name",
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "permissions": DomainObjectPropertyMetadata {
        "name": "permissions",
        "nullable": false,
        "of": {
          "nullable": false,
          "of": [
            "data:get:shipment:*",
            "data:set:shipment:*",
          ],
          "required": true,
          "type": "ENUM",
        },
        "required": true,
        "type": "ARRAY",
      },
    },
  },
]
`;

exports[`getHydratedDomainObjectMetadatasFromFiles should return metadata from files which needs hydration of nested type alias 1`] = `
[
  DomainObjectMetadata {
    "decorations": {
      "unique": [
        "onDate",
      ],
      "updatable": [],
    },
    "extends": "DomainEvent",
    "name": "PerformanceReport",
    "properties": {
      "conversions": DomainObjectPropertyMetadata {
        "name": "conversions",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
      "cost": DomainObjectPropertyMetadata {
        "name": "cost",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
      "createdAt": DomainObjectPropertyMetadata {
        "name": "createdAt",
        "nullable": false,
        "required": false,
        "type": "DATE",
      },
      "engagements": DomainObjectPropertyMetadata {
        "name": "engagements",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
      "exposures": DomainObjectPropertyMetadata {
        "name": "exposures",
        "nullable": false,
        "required": true,
        "type": "NUMBER",
      },
      "id": DomainObjectPropertyMetadata {
        "name": "id",
        "nullable": false,
        "required": false,
        "type": "NUMBER",
      },
      "onDate": DomainObjectPropertyMetadata {
        "name": "onDate",
        "nullable": false,
        "of": "StandardDate",
        "required": true,
        "type": "STRING",
      },
      "uuid": DomainObjectPropertyMetadata {
        "name": "uuid",
        "nullable": false,
        "required": false,
        "type": "STRING",
      },
    },
  },
]
`;
